src/axom/inlet/Table.hpp

```diff
@@ -19,17 +19,228 @@
 #include <functional>
 #include <unordered_map>
 #include <tuple>
+#include <type_traits>
+
+#include "fmt/fmt.hpp"
 
 #include "axom/inlet/Field.hpp"
 #include "axom/inlet/Reader.hpp"
 #include "axom/inlet/SchemaCreator.hpp"
 
 #include "axom/sidre.hpp"
 
+/*!
+ *******************************************************************************
+ * \brief Prototype for user-defined types wishing to define a by-value read
+ * from inlet with axom::inlet::Table::get(const std::string&)
+ * 
+ * This is the only way of reading in a non-default-constructible type from
+ * Inlet.
+ * 
+ * \see axom::inlet::Table::get(const std::string&)
+ *******************************************************************************
+ */
+template <typename T>
+struct FromInlet
+{
+  // "Poison" the base implementation so it cannot be used
+  // and so it can be checked in a SFINAE context
+  FromInlet() = delete;
+};
+
 namespace axom
 {
 namespace inlet
 {
+namespace detail
+{
+/*!
+ *******************************************************************************
+ * \class is_lua_primitive
+ *
+ * \brief A type trait for checking if a type is isomorphic to a Lua primitive
+ * \tparam T The type to check
+ *******************************************************************************
+ */
+template <typename T>
+struct is_lua_primitive
```

Comment:
I'm hesistant to have the input language used here.  Looking forward to future where there is going to be multiple input language options, do we see having multiples of these here (is_lua_primitive, is_json_primitive, etc) or should they be at the Reader level (LuaReader::is_primitive)?  Or should this be is_inlet_primitive?

Reply:
I don't think it's possible to implement type traits as part of an inheritance hierarchy, but for now I think it's reasonable to switch to `is_inlet_primitive` as JSON at least has roughly the same primitives.